packages/drivers/odsp-driver/src/odspDocumentStorageManager.ts

```diff
@@ -423,17 +423,29 @@ export class OdspDocumentStorageService implements IDocumentStorageService {
                     if (this.hostPolicy.concurrentSnapshotFetch && !this.hostPolicy.summarizerClient) {
                         const snapshotP = this.fetchSnapshot(hostSnapshotOptions);
 
-                        const promiseRaceWinner = await promiseRaceWithWinner([cachedSnapshotP, snapshotP]);
+                        // Ensure that failures on both paths are ignored initially.
+                        // I.e. if cache fails for some reason, we will proceed with network result.
+                        // And vice versa - if (for example) client is offline and network request fails first, we
+                        // do want to attempt to succeed with cached data!
+                        const promiseRaceWinner = await promiseRaceWithWinner([
+                            cachedSnapshotP.catch(() => undefined),
+                            snapshotP.catch(() => undefined),
+                        ]);
                         cachedSnapshot = promiseRaceWinner.value;
+                        method = promiseRaceWinner.index === 0 ? "cache" : "network";
 
                         if (cachedSnapshot === undefined) {
-                            cachedSnapshot = await snapshotP;
+                            // if network failed -> wait for cache ( then return network failure)
+                            // If cache returned empty or failed -> wait for network (success of failure)
+                            if (promiseRaceWinner.index === 1) {
+                                cachedSnapshot = await cachedSnapshotP;
+                                method = "cache";
+                            }
+                            if (cachedSnapshot === undefined) {
```

Comment:
why is this checked twice😳

Reply:
Because we are assigning value to it in the above if block. So it is checking if it is still undefined.

Reply:
oh i see! thanks!